hail/python/hail/methods/relatedness/mating_simulation.py

```diff
@@ -1,70 +1,162 @@
+
+    def get_rel(s1, s2):
+        smaller, larger = (s1, s2) if s1 < s2 else s2, s1
+        return relatedness[larger].get(smaller, 0.0)
+
+    def set_rel(from_s, to_s, value, fwd, generation_start):
+        relatedness[from_s][to_s] = value
+        if to_s >= generation_start:
+            fwd[to_s][from_s] = value
+
+    samples = [(i, f'founder_{i}', None, None) for i in range(ns)]
+    print(f'simulate_random_mating: {len(samples)} founders, {n_rounds} rounds of mating to do')
+    last_generation_start_idx = 0
+    indices = []
+
+    last_gen_fwd = defaultdict(dict)
+
+    if seed is not None:
+        random.seed(seed)
+    for generation in range(n_rounds):
+        last_generation_end = len(samples)
+        mating_generation_size = last_generation_end - last_generation_start_idx
+
+        new_pairs = int(mating_generation_size * pairs_per_generation_multiplier)
+
+        curr_gen_fwd = defaultdict(dict)
+
+        curr_sample_idx = len(samples)
+        parent_to_child = defaultdict(set)
+        for pair in range(new_pairs):
+            mother = int(random.uniform(last_generation_start_idx, last_generation_end))
+            father = int(last_generation_start_idx + (
+                    mother + random.uniform(1, mating_generation_size)) % mating_generation_size)
+
+            mother_rel = {**relatedness[mother], **last_gen_fwd[mother]}
+            father_rel = {**relatedness[father], **last_gen_fwd[father]}
+
+            merged_parent_rel = {}
+
+            for k, v in mother_rel.items():
+                merged_parent_rel[k] = .5 * v
+            for k, v in father_rel.items():
+                if k in mother_rel:
+                    merged_parent_rel[k] += .5 * v
+                else:
+                    merged_parent_rel[k] = .5 * v
+
+            if mother in merged_parent_rel:
+                assert father in merged_parent_rel
+                merged_parent_rel[mother] += 0.25
+                merged_parent_rel[father] += 0.25
+            else:
+                merged_parent_rel[mother] = 0.25
+                merged_parent_rel[father] = 0.25
+            sibling_rel = merged_parent_rel[mother]  # here mother/father should be the same, since
+
+            for child in range(children_per_pair):
+                samples.append(
+                    (curr_sample_idx, f'generation_{generation + 1}_pair_{pair}_child_{child}', mother, father))
+                for k, v in merged_parent_rel.items():
+                    set_rel(curr_sample_idx, k, v, curr_gen_fwd, last_generation_start_idx)
+
+                mother_sibs = parent_to_child[mother]
+                father_sibs = parent_to_child[father]
+                for sib in mother_sibs:
+                    if sib in father_sibs:
+                        rel = sibling_rel
+                    else:
+                        _, _, sib_mom, sib_dad = samples[sib]
+                        other_parent = sib_mom if sib_mom != mother else sib_dad
```

Comment:
AFICT, there's no sex involved here, otherwise it would obviously be the sib's dad that is the other parent, right? Since we know this sib is the mother's child.

Reply:
Correct!